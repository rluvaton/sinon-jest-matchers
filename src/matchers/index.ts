export { sinonToBeCalled } from './sinonToBeCalled';
export { sinonToBeCalledTimes } from './sinonToBeCalledTimes';
export { sinonToBeCalledWith } from './sinonToBeCalledWith';
export { sinonLastCalledWith } from './sinonLastCalledWith';
export { sinonLastReturnedWith } from './sinonLastReturnedWith';
export { sinonNthCalledWith } from './sinonNthCalledWith';
export { sinonToReturn } from './sinonToReturn';
export { sinonToReturnTimes } from './sinonToReturnTimes';
export { sinonToReturnWith } from './sinonToReturnWith';
