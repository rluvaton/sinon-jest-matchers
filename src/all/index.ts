import * as matchers from '../matchers';

// eslint-disable-next-line @typescript-eslint/ban-ts-comment
// @ts-ignore
const jestExpect = global.expect;

if (jestExpect !== undefined) {
  jestExpect.extend(matchers);
} else {
  throw new Error(
    "Unable to find Jest's global expect. " +
      'Please check you have added sinon-jest-matchers correctly to your jest configuration. ' +
      'See https://github.com/rluvaton/sinon-jest-matchers#setup for help.',
  );
}
