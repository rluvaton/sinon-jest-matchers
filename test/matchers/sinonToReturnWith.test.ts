import sinon from 'sinon';
import * as matchers from '../../src';
import * as Immutable from 'immutable';

expect.extend(matchers);

describe('sinonToReturnWith', () => {
  test('works only on stubs or sinon.spies', () => {
    const fn = function fn() {};

    expect(() => expect(fn).sinonToReturnWith(expect.anything())).toThrowErrorMatchingSnapshot();
  });

  test('works when not called', () => {
    const fn = sinon.spy();
    expect(fn).not.sinonToReturnWith('foo');

    expect(() => expect(fn).sinonToReturnWith('foo')).toThrowErrorMatchingSnapshot();
  });

  test('works with no arguments', () => {
    const fn = sinon.spy();
    fn();
    expect(fn).sinonToReturnWith();
  });

  test('works with argument that does not match', () => {
    const fn = sinon.spy(() => 'foo');
    fn();

    expect(fn).not.sinonToReturnWith('bar');

    expect(() => expect(fn).sinonToReturnWith('bar')).toThrowErrorMatchingSnapshot();
  });

  test('works with argument that does match', () => {
    const fn = sinon.spy(() => 'foo');
    fn();

    expect(fn).sinonToReturnWith('foo');

    expect(() => expect(fn).not.sinonToReturnWith('foo')).toThrowErrorMatchingSnapshot();
  });

  test('works with undefined', () => {
    const fn = sinon.spy(() => undefined);
    fn();

    expect(fn).sinonToReturnWith(undefined);

    expect(() => expect(fn).not.sinonToReturnWith(undefined)).toThrowErrorMatchingSnapshot();
  });

  test('works with Map', () => {
    const m1 = new Map([
      [1, 2],
      [2, 1],
    ]);
    const m2 = new Map([
      [1, 2],
      [2, 1],
    ]);
    const m3 = new Map([
      ['a', 'b'],
      ['b', 'a'],
    ]);

    const fn = sinon.spy(() => m1);
    fn();

    expect(fn).sinonToReturnWith(m2);
    expect(fn).not.sinonToReturnWith(m3);

    expect(() => expect(fn).not.sinonToReturnWith(m2)).toThrowErrorMatchingSnapshot();
    expect(() => expect(fn).sinonToReturnWith(m3)).toThrowErrorMatchingSnapshot();
  });

  test('works with Set', () => {
    const s1 = new Set([1, 2]);
    const s2 = new Set([1, 2]);
    const s3 = new Set([3, 4]);

    const fn = sinon.spy(() => s1);
    fn();

    expect(fn).sinonToReturnWith(s2);
    expect(fn).not.sinonToReturnWith(s3);

    expect(() => expect(fn).not.sinonToReturnWith(s2)).toThrowErrorMatchingSnapshot();
    expect(() => expect(fn).sinonToReturnWith(s3)).toThrowErrorMatchingSnapshot();
  });

  test('works with Immutable.js objects directly created', () => {
    const directlyCreated = Immutable.Map([['a', { b: 'c' }]]);
    const fn = sinon.spy(() => directlyCreated);
    fn();

    expect(fn).sinonToReturnWith(directlyCreated);

    expect(() => expect(fn).not.sinonToReturnWith(directlyCreated)).toThrowErrorMatchingSnapshot();
  });

  test('works with Immutable.js objects indirectly created', () => {
    const indirectlyCreated = Immutable.Map().set('a', { b: 'c' });
    const fn = sinon.spy(() => indirectlyCreated);
    fn();

    expect(fn).sinonToReturnWith(indirectlyCreated);

    expect(() => expect(fn).not.sinonToReturnWith(indirectlyCreated)).toThrowErrorMatchingSnapshot();
  });

  test('a call that throws is not considered to have returned', () => {
    const fn = sinon.spy(() => {
      throw new Error('Error!');
    });

    try {
      fn();
    } catch {
      // ignore error
    }

    // It doesn't matter what return value is tested if the call threw
    expect(fn).not.sinonToReturnWith('foo');
    expect(fn).not.sinonToReturnWith(null);
    expect(fn).not.sinonToReturnWith(undefined);

    expect(() => expect(fn).sinonToReturnWith(undefined)).toThrowErrorMatchingSnapshot();
  });

  // Bug in sinon - https://github.com/sinonjs/sinon/issues/2471
  test.skip('a call that throws undefined is not considered to have returned', () => {
    const fn = sinon.spy(() => {
      // eslint-disable-next-line no-throw-literal
      throw undefined;
    });

    try {
      fn();
    } catch {
      // ignore error
    }

    // It doesn't matter what return value is tested if the call threw
    expect(fn).not.sinonToReturnWith('foo');
    expect(fn).not.sinonToReturnWith(null);
    expect(fn).not.sinonToReturnWith(undefined);

    expect(() => expect(fn).sinonToReturnWith(undefined)).toThrowErrorMatchingSnapshot();
  });

  test('works with more calls than the limit', () => {
    let i = 0;
    const fn = sinon.spy(() => `foo${++i}`);

    fn();
    fn();
    fn();
    fn();
    fn();
    fn();

    expect(fn).not.sinonToReturnWith('bar');

    expect(() => {
      expect(fn).sinonToReturnWith('bar');
    }).toThrowErrorMatchingSnapshot();
  });

  test('includes the custom mock name in the error message', () => {
    const fn = sinon.spy();
    fn.named('named-mock');
    expect(fn).not.sinonToReturnWith('foo');

    expect(() => expect(fn).sinonToReturnWith('foo')).toThrowErrorMatchingSnapshot();
  });
});
