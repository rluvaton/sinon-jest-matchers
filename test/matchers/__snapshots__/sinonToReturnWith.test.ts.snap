// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sinonToReturnWith a call that throws is not considered to have returned 1`] = `
"expect(sinon.spy()).sinonToReturnWith(expected)

Expected: undefined
Received: function call threw an error

Number of returns: 0
Number of calls:   1"
`;

exports[`sinonToReturnWith includes the custom mock name in the error message 1`] = `
"expect(named-mock).sinonToReturnWith(expected)

Expected: \\"foo\\"

Number of returns: 0"
`;

exports[`sinonToReturnWith works only on stubs or sinon.spies 1`] = `
"expect(received).sinonToReturnWith(expected)

Matcher error: received value must be a stub or spy function

Received has type:  function
Received has value: [Function fn]"
`;

exports[`sinonToReturnWith works when not called 1`] = `
"expect(sinon.spy()).sinonToReturnWith(expected)

Expected: \\"foo\\"

Number of returns: 0"
`;

exports[`sinonToReturnWith works with Immutable.js objects directly created 1`] = `
"expect(sinon.spy()).not.sinonToReturnWith(expected)

Expected: not Immutable.Map {\\"a\\": {\\"b\\": \\"c\\"}}

Number of returns: 1"
`;

exports[`sinonToReturnWith works with Immutable.js objects indirectly created 1`] = `
"expect(sinon.spy()).not.sinonToReturnWith(expected)

Expected: not Immutable.Map {\\"a\\": {\\"b\\": \\"c\\"}}

Number of returns: 1"
`;

exports[`sinonToReturnWith works with Map 1`] = `
"expect(sinon.spy()).not.sinonToReturnWith(expected)

Expected: not Map {1 => 2, 2 => 1}

Number of returns: 1"
`;

exports[`sinonToReturnWith works with Map 2`] = `
"expect(sinon.spy()).sinonToReturnWith(expected)

Expected: Map {\\"a\\" => \\"b\\", \\"b\\" => \\"a\\"}
Received: Map {1 => 2, 2 => 1}

Number of returns: 1"
`;

exports[`sinonToReturnWith works with Set 1`] = `
"expect(sinon.spy()).not.sinonToReturnWith(expected)

Expected: not Set {1, 2}

Number of returns: 1"
`;

exports[`sinonToReturnWith works with Set 2`] = `
"expect(sinon.spy()).sinonToReturnWith(expected)

Expected: Set {3, 4}
Received: Set {1, 2}

Number of returns: 1"
`;

exports[`sinonToReturnWith works with argument that does match 1`] = `
"expect(sinon.spy()).not.sinonToReturnWith(expected)

Expected: not \\"foo\\"

Number of returns: 1"
`;

exports[`sinonToReturnWith works with argument that does not match 1`] = `
"expect(sinon.spy()).sinonToReturnWith(expected)

Expected: \\"bar\\"
Received: \\"foo\\"

Number of returns: 1"
`;

exports[`sinonToReturnWith works with more calls than the limit 1`] = `
"expect(sinon.spy()).sinonToReturnWith(expected)

Expected: \\"bar\\"
Received
       1: \\"foo1\\"
       2: \\"foo2\\"
       3: \\"foo3\\"

Number of returns: 6"
`;

exports[`sinonToReturnWith works with undefined 1`] = `
"expect(sinon.spy()).not.sinonToReturnWith(expected)

Expected: not undefined

Number of returns: 1"
`;
